File: c/0572-subtree-of-another-tree.c

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool isSame(struct TreeNode* p, struct TreeNode* q) {
    
    if (p == NULL && q == NULL) {
        return true;
    }
    if ((p == NULL || q == NULL) || (p -> val != q -> val)) {
        return false;
    }
    return (isSame(p -> left, q -> left) && isSame(p -> right, q -> right));
}


bool isSubtree(struct TreeNode* root, struct TreeNode* subRoot){
    
    if (root == NULL) {
        return false;
    }

    bool compareTree = false;
    if (root -> val == subRoot -> val) {
        compareTree = isSame(root, subRoot);
    }
    return (isSubtree(root -> left, subRoot) || isSubtree(root -> right, subRoot) || compareTree);
}

File: go/0572-subtree-of-another-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root == nil {
		return false
	}

	return equals(root, subRoot) || isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}

func equals(s *TreeNode, t *TreeNode) bool {
	if s == nil && t == nil {
		return true
	}

	if s == nil || t == nil || s.Val != t.Val {
		return false
	}

	return equals(s.Left, t.Left) && equals(s.Right, t.Right)
}